packages/gatsby-cli/src/create-cli.js

```diff
@@ -128,10 +135,16 @@ function buildLocalCommands(cli, isLocalSite) {
         type: `boolean`,
         default: false,
         describe: `Build site without uglifying JS bundles (for debugging).`,
+      }).option(`build-dir`, {
+        alias: `buildDirectory`,
+        type: `string`,
+        default: `public`,
+        describe: `Set build directory. Defaults to public`,
       }),
     handler: handlerP(
       getCommandHandler(`build`, (args, cmd) => {
         process.env.NODE_ENV = `production`
+        process.env.GATSBY_BUILD_DIR = args.buildDirectory || `public`
```

Comment:
We should validate that the build directory is a valid one. Someone could do `../../../../../` or whatever. Perhaps we should insist that the build directory is a simple directory name at the same level as public would be? Or would people want to do nested build directories e.g. `build1/test-1` or whatever?